tests/integration/tier0/non-destructive/kernel/test_custom_kernel.py

```diff
@@ -67,37 +57,55 @@ def custom_kernel(shell, hybrid_rocky_image):
+        available_kernel_version = shell(
+            f"{yum_call} repoquery kernel | grep -v '.src' | rpmdev-sort | tail -1"
+        ).output.strip()
+
+        # Install the kernel from the path provided by the _cross_vendor_kernel
+        # This way we don't rely on any specific version of kernel hardcoded and install what's available in the repository
+        # Disable all other repositories
+        # Call without the gpg check, so we won't need to import the GPG key
+        assert shell(f"{yum_call} install -y --nogpgcheck {available_kernel_version}").returncode == 0
+
+        custom_kernel_installed = re.sub("kernel-(?:\d+:)?", "", available_kernel_version)
+        # Assemble the full title of the custom kernel and set it as default to boot to
+        grub_substring = _get_full_kernel_title(shell, kernel=custom_kernel_installed)
+        assert shell(f"grub2-set-default '{grub_substring}'").returncode == 0
         shell("tmt-reboot -t 600")
 
     yield
 
     if os.environ["TMT_REBOOT_COUNT"] == "1":
-        # Remove the current installed kernel and install the machine default kernel.
-        custom_kernel_release = CUSTOM_KERNEL.rsplit("/", 1)[-1].replace(".rpm", "")
-        assert shell("rpm -e %s" % custom_kernel_release).returncode == 0
-
-        original_kernel = os.popen("rpm -q --last kernel | head -1 | cut -d ' ' -f1").read()
-        original_kernel_release = original_kernel.rsplit("/")[-1].replace(".rpm", "").split("-")[-1]
-
+        # This is kind of naive, but we assume the second latest installed kernel is the original one
+        # We use head to filter the first two lines of the output and tail to filter the bottom line
+        original_kernel = os.popen("rpm -q --last kernel | head -2 | tail -1 | cut -d ' ' -f1").read().strip()
+        original_kernel_title = _get_full_kernel_title(shell, kernel=original_kernel.replace("kernel-", ""))
         # Install back the CentOS 8.5 original kernel
         if "centos-8-latest" in SYSTEM_RELEASE_ENV:
-            assert shell("yum install -y %s" % original_kernel).returncode == 0
-
-        assert (
-            shell(
-                "grubby --set-default /boot/vmlinuz-*%s" % original_kernel_release,
-            ).returncode
-            == 0
-        )
+            assert shell(f"yum reinstall -y kernel").returncode == 0
+
+        grub_setup_workaround(shell)
+        assert shell(f"grub2-set-default '{original_kernel_title}'").returncode == 0
+        shell("grub2-mkconfig -o /boot/grub2/grub.cfg")
         # Reboot
         shell("tmt-reboot -t 600")
 
+    if os.environ["TMT_REBOOT_COUNT"] == "2":
+        custom_kernel_installed = os.popen("rpm -q --last kernel | head -1 | cut -d ' ' -f1").read().strip()
+        # After the system has the original kernel running, remove the custom kernel
+        kernel_to_remove = custom_kernel_installed.replace("kernel", "kernel*")
+        assert shell(f"yum remove -y {kernel_to_remove}").returncode == 0
```

Comment:
Basically, `kernel` package is just a meta-package, so removing it would not remove packages which it installed. So if I would go with just `yum remove -y kernel` it will remove just the kernel package, but I want to remove also the kernel-core, kernel-modules, etc.
```
# sudo dnf remove kernel-6.9.10-200.fc40.x86_64
Dependencies resolved.
============================================================================================================================================================================================================================
 Package                                           Architecture                                      Version                                                       Repository                                          Size
============================================================================================================================================================================================================================
Removing:
 kernel                                            x86_64                                            6.9.10-200.fc40                                               @System                                              0  

Transaction Summary
============================================================================================================================================================================================================================
Remove  1 Package

```
```
# sudo dnf remove kernel*-6.9.10-200.fc40.x86_64
Dependencies resolved.
============================================================================================================================================================================================================================
 Package                                                      Architecture                                   Version                                                  Repository                                       Size
============================================================================================================================================================================================================================
Removing:
 kernel                                                       x86_64                                         6.9.10-200.fc40                                          @System                                           0  
 kernel-core                                                  x86_64                                         6.9.10-200.fc40                                          @System                                          69 M
 kernel-modules                                               x86_64                                         6.9.10-200.fc40                                          @System                                          61 M
 kernel-modules-core                                          x86_64                                         6.9.10-200.fc40                                          @System                                          36 M
 kernel-modules-extra                                         x86_64                                         6.9.10-200.fc40                                          @System                                         2.6 M

Transaction Summary
============================================================================================================================================================================================================================
Remove  5 Packages
```
I mean working with `kernel-core` instead of just the meta-package should help as well, since removing kernel-core would remove all the packages we want to remove.
We can look into that in the future.